### PR TITLE
Add FieldVector microbenchmarks and improve fieldvector broadcast performance

### LINKED ISSUE #2067: Add FieldVector microbenchmarks and improve performance
It turns out that the `CuKernelContext` are our FieldVector kernels, since we forward those kernels to the parent of the fields. We should add microbenchmarks for these operations, and see if we can outperform cuda with some custom kernels (e.g., leverage linear indexing).

Screenshot 2024-11-01 at 10 16 19 AM

### PR DESCRIPTION
**Pull request by charleskawczynski**
Closes #2067. This turned out to be pretty easy: any fieldvector operation is embarrassingly parallel. We can leverage `NonExtrudedBroadcasted`, forward everything to the backing arrays, and just linearly index everywhere. This puts us in a pretty good shape for fieldvector operations, and it'll be even better at lower resolution since we parallelize across field variables.

#### main
```julia
N reads-writes: 5,  Float_type = Float64, Device_bandwidth_GBs=2039
┌─────────────┬───────────────────────────────────┬─────────┬─────────────┬──────────────┬────────┐
│ funcs       │ time per call                     │ bw %    │ achieved bw │ problem size │ n-reps │
├─────────────┼───────────────────────────────────┼─────────┼─────────────┼──────────────┼────────┤
│ FieldVector │ 1 millisecond, 104 microseconds   │ 54.161  │ 1104.34     │ (32745600,)  │ 4475   │
│ FieldVector │ 307 microseconds, 597 nanoseconds │ 48.6243 │ 991.45      │ (8186400,)   │ 10000  │
└─────────────┴───────────────────────────────────┴─────────┴─────────────┴──────────────┴────────┘
```
#### this PR
```julia
N reads-writes: 5,  Float_type = Float64, Device_bandwidth_GBs=2039
┌─────────────┬───────────────────────────────────┬─────────┬─────────────┬──────────────┬────────┐
│ funcs       │ time per call                     │ bw %    │ achieved bw │ problem size │ n-reps │
├─────────────┼───────────────────────────────────┼─────────┼─────────────┼──────────────┼────────┤
│ FieldVector │ 757 microseconds, 513 nanoseconds │ 78.9779 │ 1610.36     │ (32745600,)  │ 6491   │
│ FieldVector │ 202 microseconds, 849 nanoseconds │ 73.7335 │ 1503.43     │ (8186400,)   │ 10000  │
└─────────────┴───────────────────────────────────┴─────────┴─────────────┴──────────────┴────────┘
```

#### main
```julia
N reads-writes: 5,  Float_type = Float64,
┌─────────────┬──────────────────────────────────┬─────────────┬──────────────┬────────┐
│ funcs       │ time per call                    │ achieved bw │ problem size │ n-reps │
├─────────────┼──────────────────────────────────┼─────────────┼──────────────┼────────┤
│ FieldVector │ 32 milliseconds, 45 microseconds │ 38.0674     │ (32745600,)  │ 33     │
│ FieldVector │ 7 milliseconds, 792 microseconds │ 39.1372     │ (8186400,)   │ 529    │
└─────────────┴──────────────────────────────────┴─────────────┴──────────────┴────────┘
```

#### This PR

```julia
N reads-writes: 5,  Float_type = Float64,
┌─────────────┬───────────────────────────────────┬─────────────┬──────────────┬────────┐
│ funcs       │ time per call                     │ achieved bw │ problem size │ n-reps │
├─────────────┼───────────────────────────────────┼─────────────┼──────────────┼────────┤
│ FieldVector │ 23 milliseconds, 218 microseconds │ 52.5398     │ (32745600,)  │ 43     │
│ FieldVector │ 5 milliseconds, 752 microseconds  │ 53.0143     │ (8186400,)   │ 828    │
└─────────────┴───────────────────────────────────┴─────────────┴──────────────┴────────┘
```

### CODE REVIEW

File: src/Fields/fieldvector.jl [hunk 1]
```diff
@@ -218,6 +209,20 @@ function Spaces.weighted_dss!(
...
+    Base.Broadcast.broadcasted(fs, RecursiveApply.:⊟, args...)
+
+Base.Broadcast.broadcasted(fs::FieldVectorStyle, ::typeof(*), args...) =
+    Base.Broadcast.broadcasted(fs, RecursiveApply.:⊠, args...)
+
+Base.Broadcast.broadcasted(fs::FieldVectorStyle, ::typeof(/), args...) =
+    Base.Broadcast.broadcasted(fs, RecursiveApply.rdiv, args...)
+
+Base.Broadcast.broadcasted(fs::FieldVectorStyle, ::typeof(muladd), args...) =
+    Base.Broadcast.broadcasted(fs, RecursiveApply.rmuladd, args...)
```
Comment (charleskawczynski) [inline src/Fields/fieldvector.jl:225]: We might not actually need this

File: src/Fields/fieldvector.jl [hunk 2]
```diff
@@ -332,24 +322,41 @@ end
...
+
+@inline function Base.copyto!(
+    dest::FieldVector,
+    bc::Base.Broadcast.Broadcasted{Style},
+) where {
+    Style <:
+    Union{Base.Broadcast.AbstractArrayStyle{0}, Base.Broadcast.Style{Tuple}},
+}
+    map(propertynames(dest)) do symb
+        Base.@_inline_meta
```
Comment (Sbozzolo) [inline src/Fields/fieldvector.jl:344]: Is this inlining the anonymous function for the map?

**Review by Sbozzolo — APPROVED** [review]

File: src/Fields/fieldvector.jl (hunk 2 above)
Comment (charleskawczynski) [reply src/Fields/fieldvector.jl:344]: Yes

**Comment by charleskawczynski** [comment]
Interesting, this shows that ClimaLand is using fieldvectors in some untested edge case. I'm going to find out the difference and add some unit tests that exercise this.

**Comment by charleskawczynski** [comment]
This looks done! 🎉

**Comment by charleskawczynski** [comment]
Just to comment on why this took a bit longer to iron out: the land model seems to rely on recursively defined fieldvectors, which we support `copyto!` for. Doing this correctly is a bit tricky because we need to dispatch to a custom (and device-specific) method to better optimize. All that said, I'm happy with the result.
